src/gateway_api/client/gateway_client.rs

```diff
@@ -0,0 +1,216 @@
+        assert_eq!(
+            result,
+            Err(CommonError::NetworkRequestInvalidUrl { bad_value })
+        )
+    }
+
+    #[actix_rt::test]
+    async fn execute_network_request_bad_status_code() {
+        let mock_antenna = MockAntenna::new(
+            404, // bad code
+            (),
+        );
+        let sut = SUT::new(Arc::new(mock_antenna), NetworkID::Stokenet);
+        let req = sut.current_epoch();
+        let result = timeout(MAX, req).await.unwrap();
+        assert_eq!(result, Err(CommonError::NetworkResponseBadCode))
+    }
+
+    #[actix_rt::test]
+    async fn execute_network_request_empty_body() {
+        let mock_antenna = MockAntenna::new(
+            200,
+            (), // empty body
+        );
+        let sut = SUT::new(Arc::new(mock_antenna), NetworkID::Stokenet);
+        let req = sut.current_epoch();
+        let result = timeout(MAX, req).await.unwrap();
+        assert_eq!(result, Err(CommonError::NetworkResponseEmptyBody))
+    }
+
+    #[actix_rt::test]
+    async fn execute_network_request_invalid_json() {
+        let mock_antenna = MockAntenna::new(
+            200,
+            BagOfBytes::sample_aced(), // wrong JSON
+        );
+        let sut = SUT::new(Arc::new(mock_antenna), NetworkID::Stokenet);
+        let req = sut.current_epoch();
+        let result = timeout(MAX, req).await.unwrap();
+        assert_eq!(result, Err(CommonError::NetworkResponseJSONDeserialize { into_type: "sargon::gateway_api::models::types::response::transaction::construction::transaction_construction_response::TransactionConstructionResponse".to_owned() }))
+    }
+
+    #[actix_rt::test]
+    async fn spy_headers() {
+        let mock_antenna = MockAntenna::with_spy(200, (), |request| {
+            assert_eq!(
+                request
+                    .headers
+                    .keys()
+                    .into_iter()
+                    .map(|v| v.to_string())
+                    .collect::<BTreeSet<String>>(),
+                [
+                    "RDX-Client-Version",
+                    "RDX-Client-Name",
+                    "accept",
+                    "content-Type",
+                    "user-agent"
+                ]
+                .into_iter()
+                .map(|s| s.to_owned())
+                .collect::<BTreeSet<String>>()
+            )
+        });
+        let sut = SUT::new(Arc::new(mock_antenna), NetworkID::Stokenet);
+        let req = sut.current_epoch();
+        drop(timeout(MAX, req).await.unwrap());
+    }
+
+    #[actix_rt::test]
+    async fn test_submit_notarized_transaction_mock_duplicate() {
+        let mock_antenna =
+            MockAntenna::with_response(TransactionSubmitResponse {
+                duplicate: true,
+            });
+        let sut = SUT::new(Arc::new(mock_antenna), NetworkID::Stokenet);
+        let req =
+            sut.submit_notarized_transaction(NotarizedTransaction::sample());
```

Comment:
testing new method, by mocking that it is a duplicate and assert we get `CommonError::GatewaySubmitDuplicateTX`